### ISSUE TITLE
snapshotbackpopulate controller handler triggers every second

### ISSUE THREAD

**Issue by superseb** [rancher/rancher#32719]
**What kind of request is this (question/bug/enhancement/feature request):**
bug

**Steps to reproduce (least amount of steps as possible):**
Run Rancher with loglevel `trace` and add a custom cluster with a single node with all roles and wait for provisioning to complete.

**Result:**
See logs flooded with `snapshotbackpopulate` handler:

```
2021/05/13 19:17:59 [TRACE] ConfigMapController calling handler snapshotbackpopulate kube-system/cattle-controllers
2021/05/13 19:18:00 [TRACE] ConfigMapController calling handler snapshotbackpopulate kube-system/cattle-controllers
2021/05/13 19:18:01 [TRACE] ConfigMapController calling handler snapshotbackpopulate kube-system/k3s
2021/05/13 19:18:01 [TRACE] ConfigMapController calling handler snapshotbackpopulate kube-system/cattle-controllers
2021/05/13 19:18:02 [TRACE] ConfigMapController calling handler snapshotbackpopulate kube-system/cattle-controllers
2021/05/13 19:18:03 [TRACE] ConfigMapController calling handler snapshotbackpopulate kube-system/k3s
2021/05/13 19:18:03 [TRACE] ConfigMapController calling handler snapshotbackpopulate kube-system/cattle-controllers
```

This is because the handler is set on `kube-system` ConfigMap changes and `cattle-controllers` is used for leader election and triggers the handler. Introduced by https://github.com/rancher/rancher/blame/91a6f51f259acd5670ff167d6bcf1916da7406f1/pkg/controllers/managementuser/snapshotbackpopulate/snapshotbackpopulate.go#L37

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): `master-c0508344405fb9bd55198d2196e93fe3fc33ade7-head`
- Installation option (single install/HA): single

**Comment by Jono-SUSE-Rancher** [comment]
@kinarashah is this related to the controller issues we are seeing with the management agent?

**Comment by snasovich** [comment]
@superseb , took initial look into it. Have some thoughts to share and could use some guidance.
As you listed in the description, `snapshotbackpopulate` controller is subscribed to any changes to any config map in `kube-system` namespace while in fact it's only interested in changes to `k3s-etcd-snapshots` and `rke2-etcd-snapshots` config maps. It doesn't do anything for other config maps, but since logging comes from Norman's genericController it's produced for any config map change, including `cattle-controllers` that is "updated" as part of leader election mechanism ("updated" is because I believe it's not actually updated, but nevertheless `resourceVersion` gets bumped up which triggers update).

Now, ideally we would like to change this controller so it's triggered by updates to these specific config maps and not just any config map. We could also look into reducing frequency of updates on `cattle-controllers` config map, but it won't be a complete fix of the issue anyway and would require core changes.

First, I'm not finding any existing controllers that subscribe to changes to specific config map. In fact this is the only controller I see that uses `ConfigMapInterface.AddHandler` (or `ConfigMapInterface.Controller().AddHandler`). I see we can query specific config map by using `ConfigMapInterface.Get`, but the returned `ConfigMap` object doesn't seem to feature a way to subscribe to changes.
Second, I'm not sure that when `Register` is called all config maps are already available and can be located.
Also, we need to cover for the case when these config maps we're interested in do not exist initially and are created at later point.

Appreciate any information / suggestions you can provide to help me proceed with this issue.

P.S. Maybe there is a way to create instance of `ConfigMapInterface` that watches not just by namespace, but also by specified "selector" func. Just started looking this way, but this seems to require changes in Norman. Anyway, at this point looking for some guidance on an approach for addressing this issue.

**Comment by snasovich** [comment]
Looking at the following lines it looks like `genericController` is invoked on change to any Config Map regardless of the namespace:
https://github.com/rancher/norman/blob/59b3523c3133434a3943f05cbd8cc6409ea976d5/controller/generic_controller.go#L55-L60
So `genericController` has built-in "by namespace" filtering capabilities. We could extend these capabilities further by allowing to pass func that determines whether the handler is applicable to a given `runtime.Object` (so we can move check from https://github.com/rancher/rancher/blob/cdc12d8a5cc6fda069709a950f9a6dc9e38892e1/pkg/controllers/managementuser/snapshotbackpopulate/snapshotbackpopulate.go#L45 to func passed to `AddHandler` as an additional argument). However, adopting this approach won't really change the fact the `genericController` is still invoked on change to any objects of a given kind (Config Map in this case), so it doesn't offer noticeable ROI for something that requires framework (`rancher/norman`) changes.
Also, when https://github.com/rancher/rancher/issues/33140 is implemented, the Controllers that are subscribed to Config Map changes won't be triggered so often.

All this said, I'm proposing to leave this behavior as "functions as designed".

@superseb @ibuildthecloud , please let me know your thoughts on this.

**Comment by snasovich** [comment]
PR: https://github.com/rancher/rancher/pull/33377

**Comment by snasovich** [comment]
@sowmyav27 , this is now in `master-head` and ready to test.

**Comment by sowmyav27** [comment]
To validate:
- log messages should not fire every second or so (after Rancher “settles”)

**Comment by snasovich** [comment]
@thedadams , please validate this one. Thank you.

**Comment by thedadams** [comment]
The `snapshotbackpopulate` controller seems to be broken at this time. The logs I see now are:
```
2021/08/06 21:05:46 [ERROR] error syncing 'kube-system/rke2-etcd-snapshots': handler snapshotbackpopulate: Index with name by-cluster does not exist, requeuing
2021/08/06 21:07:25 [ERROR] error syncing 'kube-system/rke2-etcd-snapshots': handler snapshotbackpopulate: Index with name by-cluster does not exist, requeuing
2021/08/06 21:07:46 [ERROR] error syncing 'kube-system/rke2-etcd-snapshots': handler snapshotbackpopulate: Index with name by-cluster does not exist, requeuing
2021/08/06 21:09:25 [ERROR] error syncing 'kube-system/rke2-etcd-snapshots': handler snapshotbackpopulate: Index with name by-cluster does not exist, requeuing
2021/08/06 21:09:46 [ERROR] error syncing 'kube-system/rke2-etcd-snapshots': handler snapshotbackpopulate: Index with name by-cluster does not exist, requeuing
2021/08/06 21:11:25 [ERROR] error syncing 'kube-system/rke2-etcd-snapshots': handler snapshotbackpopulate: Index with name by-cluster does not exist, requeuing
2021/08/06 21:11:46 [ERROR] error syncing 'kube-system/rke2-etcd-snapshots': handler snapshotbackpopulate: Index with name by-cluster does not exist, requeuing
```

**Comment by thedadams** [comment]
In order to test this, it should be assured that the `TRACE` messages in the original description do not appear every second.

In addition, it should be assured that the error messages above do not appear.

**Comment by deniseschannon** [comment]
@SheilaghM @snasovich At this point, I'm not sure who is the person who is dev validating this issue as both @superseb and @thedadams have put in efforts to fix the reported issue.

**Comment by snasovich** [comment]
@deniseschannon , the fix was actually done by @ibuildthecloud . So having @superseb do validation seems to be OK and actually makes most sense since he's the one who reported the issue originally.

**Comment by SheilaghM** [comment]
I asked @superseb to validate.  I can't unassign @thedadams.

**Comment by snasovich** [comment]
@SheilaghM , thank you. It's OK to keep Donnie assigned. In fact for history/filtering purposes it's better to leave original developer assigned.

**Comment by superseb** [comment]
Reproduced using:

```
$ docker run -d --privileged --name rancher-server rancher/rancher:master-c0508344405fb9bd55198d2196e93fe3fc33ade7-head --trace >/dev/null 2>&1 && sleep 150 && docker logs rancher-server 2>&1 | grep snapshotbackpopulate | wc -l
     107
```

Validated fixed using `v2.6.0-rc1` and `v2.6-head` (`723f747b1`):

```
$ docker run -d --privileged --name rancher-server rancher/rancher:v2.6.0-rc1 --trace >/dev/null 2>&1 && sleep 150 && docker logs rancher-server 2>&1 | grep snapshotbackpopulate | wc -l
       0
```

```
$ docker run -d --privileged --name rancher-server rancher/rancher:v2.6-head --trace >/dev/null 2>&1 && sleep 150 && docker logs rancher-server 2>&1 | grep snapshotbackpopulate | wc -l
       0
```

**Comment by snasovich** [comment]
@sowmyav27 , please take a look at validation above and close if it looks OK to you.
